File: backend/internal/platform/ids/ids.go
// Package ids formats and parses the SWP-<ENTITY>-<NUMERIC> identifiers
// (CONVENTIONS §4). IDs are OPAQUE to clients; only the server constructs them.
// The numeric portion is allocated per-prefix by the `swp_next_id()` Postgres
// function (see migrations) — repositories call the NextID query inside the
// same transaction as the insert, then Format the result.
package ids

import (
	"fmt"
	"strconv"
	"strings"
)

// Prefix is the per-entity ID prefix (CONVENTIONS §4 entity table).
type Prefix string

const (
	User                Prefix = "USR"
	AuditLog            Prefix = "AL"
	Employee            Prefix = "EMP"
	EmploymentAgreement Prefix = "AG"
	ClientCompany       Prefix = "CMP"
	ClientSite          Prefix = "SITE"
	ServiceLine         Prefix = "SVC"
	Position            Prefix = "POS"
	LeaveType           Prefix = "LT"
	AttendanceCode      Prefix = "AC"
	OvertimeRule        Prefix = "OTR"
	ChangeRequest       Prefix = "CHG"
	Placement           Prefix = "PL"
	ShiftLeaderAssign   Prefix = "SLA"
	ShiftMaster         Prefix = "SHF"
	ScheduleEntry       Prefix = "SCH"
	Attendance          Prefix = "ATT"
	AttendanceCorrect   Prefix = "COR"
	LeaveRequest        Prefix = "LR"
	LeaveQuota          Prefix = "LQ"
	Overtime            Prefix = "OT"
	PublicHoliday       Prefix = "HOL"
	Payslip             Prefix = "PS"
	Notification        Prefix = "NTF"
	ExportJob           Prefix = "EXP"
)

const namespace = "SWP"

// Format builds an opaque ID, e.g. Format(Employee, 1042) -> "SWP-EMP-1042".
func Format(p Prefix, n int64) string {
	return fmt.Sprintf("%s-%s-%d", namespace, p, n)
}

// Parse splits an ID back into prefix + numeric. Used internally only (e.g. to
// pick the right sequence); never expose the numeric to clients.
func Parse(id string) (Prefix, int64, error) {
	parts := strings.SplitN(id, "-", 3)
	if len(parts) != 3 || parts[0] != namespace {
		return "", 0, fmt.Errorf("malformed id %q", id)
	}
	n, err := strconv.ParseInt(parts[2], 10, 64)
	if err != nil {
		return "", 0, fmt.Errorf("malformed id numeric %q: %w", id, err)
	}
	return Prefix(parts[1]), n, nil
}

File: frontend/packages/shared/src/ids.ts
/**
 * Branded SWP-<ENTITY>-<NUMERIC> ID types. ENGINEERING.md A3 · CONVENTIONS.md §4.
 *
 * IDs are OPAQUE strings — never parse the numeric portion. Branding stops an
 * `SWP-EMP` id being passed where an `SWP-PL` is expected, at compile time.
 */

declare const __brand: unique symbol;
type Brand<T extends string> = { readonly [__brand]: T };

/** A prefixed, branded id string. */
export type Id<P extends string> = string & Brand<P>;

// Entity prefixes — CONVENTIONS.md §4 entity-prefix table (keep in sync).
export const ID_PREFIX = {
  USER: 'SWP-USR',
  AUDIT_LOG: 'SWP-AL',
  EMPLOYEE: 'SWP-EMP',
  AGREEMENT: 'SWP-AG',
  CLIENT_COMPANY: 'SWP-CMP',
  LEAVE_TYPE: 'SWP-LT',
  ATTENDANCE_CODE: 'SWP-AC',
  OVERTIME_RULE: 'SWP-OTR',
  CHANGE_REQUEST: 'SWP-CHG',
  PLACEMENT: 'SWP-PL',
  SHIFT_LEADER_ASSIGNMENT: 'SWP-SLA',
  SHIFT_MASTER: 'SWP-SHF',
  SCHEDULE: 'SWP-SCH',
  ATTENDANCE: 'SWP-ATT',
  CORRECTION: 'SWP-COR',
  LEAVE_REQUEST: 'SWP-LR',
  LEAVE_QUOTA: 'SWP-LQ',
  OVERTIME: 'SWP-OT',
  HOLIDAY: 'SWP-HOL',
  PAYSLIP: 'SWP-PS',
  NOTIFICATION: 'SWP-NTF',
  EXPORT: 'SWP-EXP',
} as const;

export type EntityPrefix = (typeof ID_PREFIX)[keyof typeof ID_PREFIX];

// Convenience aliases used across the app.
export type UserId = Id<typeof ID_PREFIX.USER>;
export type EmployeeId = Id<typeof ID_PREFIX.EMPLOYEE>;
export type ClientCompanyId = Id<typeof ID_PREFIX.CLIENT_COMPANY>;
export type PlacementId = Id<typeof ID_PREFIX.PLACEMENT>;
export type ScheduleId = Id<typeof ID_PREFIX.SCHEDULE>;
export type AttendanceId = Id<typeof ID_PREFIX.ATTENDANCE>;
export type LeaveRequestId = Id<typeof ID_PREFIX.LEAVE_REQUEST>;
export type OvertimeId = Id<typeof ID_PREFIX.OVERTIME>;

const ID_RE = /^SWP-[A-Z]+-\d+$/;

/** True if `value` is a structurally valid SWP id with the given prefix. */
export function isId<P extends EntityPrefix>(prefix: P, value: unknown): value is Id<P> {
  return typeof value === 'string' && value.startsWith(`${prefix}-`) && ID_RE.test(value);
}

/** Assert + brand. Throws on malformed/wrong-prefix input (use at trust boundaries). */
export function asId<P extends EntityPrefix>(prefix: P, value: string): Id<P> {
  if (!isId(prefix, value)) {
    throw new Error(`Invalid id: expected ${prefix}-<n>, got "${value}"`);
  }
  return value;
}

/** Extract the entity prefix from any SWP id (for display/routing), or null if malformed. */
export function prefixOf(value: string): string | null {
  const m = value.match(/^(SWP-[A-Z]+)-\d+$/);
  return m?.[1] ?? null;
}
